### ISSUE TITLE
FileHttpRequestMessageFactory.PutRange nullref on Linux

### ISSUE THREAD

**Issue by chlowell** [Azure/azure-storage-net#225]
Using 6.2.2-preview with CoreCLR runtime 1.0.0-rc1-update1 on Linux, I always see a nullref when I call CloudFile.UploadFromFileAsync.

Here's a repro project:

### Program.cs
```cs
using System.IO;
using Microsoft.WindowsAzure.Storage;
using Microsoft.WindowsAzure.Storage.File;

public static class Program
{
    public static void Main(string[] args)
    {
        var connectionString = "your secrets here";
        var fileToUpload = "Program.cs";

        CloudStorageAccount storageAccount = CloudStorageAccount.Parse(connectionString);
        CloudFileClient fileClient = storageAccount.CreateCloudFileClient();
        CloudFileShare share = fileClient.GetShareReference("reprofileshare");
        share.CreateIfNotExistsAsync().GetAwaiter().GetResult();
        CloudFileDirectory root = share.GetRootDirectoryReference();
        CloudFileDirectory dir = root.GetDirectoryReference("reprofiledirectory");
        dir.CreateIfNotExistsAsync().GetAwaiter().GetResult();
        CloudFile file = dir.GetFileReference(fileToUpload);
        file.UploadFromFileAsync(fileToUpload, FileMode.Open).GetAwaiter().GetResult();
        // NullReferenceException
    }
}
```
### project.json
```json
{
  "version": "1.0.0-*",
  "dependencies": {
    "WindowsAzure.Storage": "6.2.2-preview"
  },
  "frameworks": {
    "dnxcore50": {
      "dependencies": {
        "System.IO.FileSystem": "4.0.1-beta-23516"
      }
    }
  },
  "compile": [
    "**/*.cs"
  ]
}
```

### output
```
$ dnx run
Microsoft.WindowsAzure.Storage.WrappedStorageException (0x80004003): <?xml version="1.0" encoding="utf-16"?>
<!--An exception has occurred. For more information please deserialize this message via RequestResult.TranslateFromExceptionMessage.-->
<RequestResult>
  <HTTPStatusCode>0</HTTPStatusCode>
  <HttpStatusMessage />
  <TargetLocation>Primary</TargetLocation>
  <ServiceRequestID />
  <ContentMd5 />
  <Etag />
  <RequestDate />
  <StartTime>Tue, 12 Jan 2016 18:04:25 GMT</StartTime>
  <EndTime>Mon, 01 Jan 0001 00:00:00 GMT</EndTime>
  <Error></Error>
  <ExceptionInfo>
    <Type>StorageException</Type>
    <HResult>-2146233088</HResult>
    <Message>Object reference not set to an instance of an object.</Message>
    <Source />
    <StackTrace />
    <InnerExceptionInfo>
      <ExceptionInfo>
        <Type>NullReferenceException</Type>
        <HResult>-2147467261</HResult>
        <Message>Object reference not set to an instance of an object.</Message>
        <Source>Microsoft.WindowsAzure.Storage</Source>
        <StackTrace>   at Microsoft.WindowsAzure.Storage.File.Protocol.FileHttpRequestMessageFactory.PutRange(Uri uri, Nullable`1 timeout, FileRange fileRange, FileRangeWrite fileRangeWrite, AccessCondition accessCondition, HttpContent content, OperationContext operationContext)
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.&lt;ExecuteAsyncInternal&gt;d__4`1.MoveNext()</StackTrace>
      </ExceptionInfo>
    </InnerExceptionInfo>
  </ExceptionInfo>
</RequestResult> ---> Microsoft.WindowsAzure.Storage.StorageException: Object reference not set to an instance of an object. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.WindowsAzure.Storage.File.Protocol.FileHttpRequestMessageFactory.PutRange(Uri uri, Nullable`1 timeout, FileRange fileRange, FileRangeWrite fileRangeWrite, AccessCondition accessCondition, HttpContent content, OperationContext operationContext)
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.<ExecuteAsyncInternal>d__4`1.MoveNext()
   --- End of inner exception stack trace ---
Request Information
RequestID:
RequestDate:
StatusMessage:

   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.<ExecuteAsyncInternal>d__4`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.WindowsAzure.Storage.File.CloudFile.<>c__DisplayClass115_0.<<WriteRangeAsync>b__0>d.MoveNext()
Microsoft.WindowsAzure.Storage.StorageException: Object reference not set to an instance of an object. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.WindowsAzure.Storage.File.Protocol.FileHttpRequestMessageFactory.PutRange(Uri uri, Nullable`1 timeout, FileRange fileRange, FileRangeWrite fileRangeWrite, AccessCondition accessCondition, HttpContent content, OperationContext operationContext)
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.<ExecuteAsyncInternal>d__4`1.MoveNext()
   --- End of inner exception stack trace ---
Request Information
RequestID:
RequestDate:
StatusMessage:

System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.WindowsAzure.Storage.File.Protocol.FileHttpRequestMessageFactory.PutRange(Uri uri, Nullable`1 timeout, FileRange fileRange, FileRangeWrite fileRangeWrite, AccessCondition accessCondition, HttpContent content, OperationContext operationContext)
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.<ExecuteAsyncInternal>d__4`1.MoveNext()
```

**Comment by pemari-msft** [comment]
This seems to be a CoreCLR/Linux specific interaction. As an aside, we haven't supported `update1` from CoreCLR yet but thanks for providing such a good repro. We'll take a look and see if there's something we can do as a workaround.

**Comment by chlowell** [comment]
This occurs on ```final``` as well. Which is the latest supported version?

**Comment by pemari-msft** [comment]
[rc1](https://github.com/Azure/azure-storage-net#target-frameworks). Not that I think that's the issue. Like I said, I think it's something to do with CoreCLR on Linux as it works on Windows.

**Comment by pemari-msft** [comment]
Hi @chlowell, 

We've determined that there's a bug in the CoreCLR JIT code on Linux. This causes the nullref on that you're seeing and makes sense as to why it's platform specific. We've done some dissecting with the relevant teams and they are aiming to get a fix to this in by `1.0.0-rc2`. Sorry for the inconvenience. A temporary workaround for now is to use a debug build of our library (the debug build skips this particular optimization in the JIT). 

Thanks for the bug report!
